components/eventing-controller/pkg/handlers/nats_test.go

```diff
@@ -87,29 +89,93 @@ func TestConvertMsgToCE(t *testing.T) {
-	return jsonCE
-}
+func TestSubscription(t *testing.T) {
+	natsPort := 5222
+	subscriberPort := 8080
+	subscriberReceiveURL := fmt.Sprintf("http://127.0.0.1:%d/subscribe", subscriberPort)
+	subscriberCheckURL := fmt.Sprintf("http://127.0.0.1:%d/check", subscriberPort)
 
-func NewCloudEvent(data, id, source, eventTime, eventType string, t *testing.T) cev2event.Event {
-	timeInRFC3339, err := time.Parse(time.RFC3339, eventTime)
+	// Start Nats server
+	natsServer := eventingtesting.RunNatsServerOnPort(natsPort)
+	defer natsServer.Shutdown()
+
+	natsURL := natsServer.ClientURL()
+	natsClient := Nats{
+		Subscriptions: make(map[string]*nats.Subscription),
+		Config: env.NatsConfig{
+			Url:           natsURL,
+			MaxReconnects: 2,
+			ReconnectWait: time.Second,
+		},
+		Log: ctrl.Log.WithName("reconciler").WithName("Subscription"),
+	}
+	err := natsClient.Initialize()
 	if err != nil {
-		t.Fatalf("failed to parse time: %v", err)
+		t.Fatalf("failed to connect to Nats Server: %v", err)
 	}
-	dataContentType := "application/json"
-	return cev2event.Event{
-		Context: &cev2event.EventContextV1{
-			Type: eventType,
-			Source: types.URIRef{
-				URL: url.URL{
-					Path: source,
-				},
-			},
-			ID:              id,
-			DataContentType: &dataContentType,
-			Time:            &types.Timestamp{Time: timeInRFC3339},
-		},
-		DataEncoded: []byte(data),
+
+	// Create a new subscriber
+	subscriber := eventingtesting.NewSubscriber(fmt.Sprintf(":%d", subscriberPort), t)
+	subscriber.Start()
+
+	// Shutting down subscriber
+	defer subscriber.Shutdown()
+
+	// Check subscriber is running or not by checking the store
+	err = subscriber.CheckEvent("", subscriberCheckURL)
+	if err != nil {
+		t.Errorf("subscriber did not receive the event: %v", err)
+	}
+
+	// Create a subscription
+	sub := eventingtesting.NewSubscription("sub", "foo", eventingtesting.WithFilterForNats)
+	sub.Spec.Sink = subscriberReceiveURL
+	err = natsClient.SyncSubscription(sub)
+	if err != nil {
+		t.Fatalf("failed to Sync subscription: %v", err)
+	}
+
+	data := "sampledata"
+	// Send an event
+	err = SendEvent(&natsClient, data)
+	if err != nil {
+		t.Fatalf("failed to publish event: %v", err)
 	}
+
+	// Check for the event
+	err = subscriber.CheckEvent(data, subscriberCheckURL)
+	if err != nil {
+		t.Errorf("subscriber did not receive the event: %v", err)
+	}
+
+	// Delete subscription
+	err = natsClient.DeleteSubscription(sub)
+	if err != nil {
+		t.Errorf("failed to delete subscription: %v", err)
+	}
+
+	newData := "newdata"
+	// Send an event
+	err = SendEvent(&natsClient, newData)
+	if err != nil {
+		t.Fatalf("failed to publish event: %v", err)
+	}
+
+	// Check for the event that it did not reach subscriber
+	err = subscriber.CheckEvent(newData, subscriberCheckURL)
+	if err == nil {
+		t.Error("subscription still exists in Nats")
```

Comment:
The error message is confusing to me because the reasons/return values can be different. at least we don't check if the string is still `sampledata` from the last subscription for example.

Reply:
True, I refactored to handle the err based on the message.